compiler/rustc_data_structures/src/carbon_hash.rs

```diff
@@ -0,0 +1,267 @@
+use std::hash::{Hash, Hasher};
+
+#[inline]
+pub fn hash_value<T: Hash>(value: &T) -> u64 {
+    let mut buf = CarbonHasher::new();
+    value.hash(&mut buf);
+    buf.hash_code()
+}
+
+#[inline]
+pub fn hash_value_with_seed<T: Hash>(value: &T, seed: u64) -> u64 {
+    let mut buf = CarbonHasher::new_with_seed(seed);
+    value.hash(&mut buf);
+    buf.hash_code()
+}
+
+#[derive(Default)]
+pub struct CarbonHasher {
+    buffer: u64,
+}
+
+const MUL_CONSTANT: u64 = 0x9e37_79b9_7f4a_7c15;
+
+impl CarbonHasher {
+    #[inline]
+    pub fn new_with_seed(seed: u64) -> Self {
+        Self { buffer: seed }
+    }
+
+    #[inline]
+    pub fn new() -> Self {
+        Self { buffer: 0 }
+    }
+
+    #[inline]
+    pub fn hash_code(self) -> u64 {
+        self.buffer
+    }
+
+    #[inline]
+    pub fn hash_sized_bytes(&mut self, bytes: &[u8]) {
+        let data_ptr = bytes.as_ptr();
+        let size = bytes.len();
+
+        if core::intrinsics::unlikely(size == 0) {
+            self.hash_one(0);
+            return;
+        }
+        if size <= 8 {
+            let data = if size >= 4 {
+                unsafe { read_4_to_8(data_ptr, size) }
+            } else {
+                unsafe { read_1_to_3(data_ptr, size) }
+            };
+
+            self.buffer = mix(data ^ self.buffer, sample_random_data(size));
+            return;
+        }
+        if size <= 16 {
+            let data = unsafe { read_8_to_16(data_ptr, size) };
+            self.buffer = mix(data.0 ^ sample_random_data(size), data.1 ^ self.buffer);
+            return;
+        }
+        if size <= 32 {
+            self.buffer ^= sample_random_data(size);
+            let m0 = mix(
+                unsafe { read_8(data_ptr) } ^ STATIC_RANDOM_DATA[1],
+                unsafe { read_8(data_ptr.add(8)) } ^ self.buffer,
+            );
+            let tail_16b_ptr = unsafe { data_ptr.add(size - 16) };
+            let m1 = mix(
+                unsafe { read_8(tail_16b_ptr) } ^ STATIC_RANDOM_DATA[3],
+                unsafe { read_8(tail_16b_ptr.add(8)) } ^ self.buffer,
+            );
+
+            self.buffer = m0 ^ m1;
+            return;
+        }
+
+        self.hash_bytes_large(bytes);
+    }
+
+    #[inline]
```

Comment:
I'd encourage this to not be inline FWIW. The code is quite large and doesn't seem to benefit much from it.